activedata_etl/transforms/jscov_to_es.py

```diff
@@ -13,61 +13,120 @@
-        response_stream = http.get(artifact.url).raw
-        with Timer("Processing {{jscov_file}}", param={"jscov_file": artifact.url}):
+        with ZipFile(jsdcov_file) as zipped:
+            for num, zip_name in enumerate(zipped.namelist()):
+                json_stream = ibytes2ilines(zipped.open(zip_name))
+                counter = count_generator().next
+                for source_file_index, obj in enumerate(stream.parse(json_stream, '.', ['.'])):
+                    if please_stop:
+                        Log.error("Shutdown detected. Stopping job ETL.")
+
+                    if source_file_index == 0:
+                        # this is not a jscov object but an object containing the version metadata
+                        # TODO: this metadata should not be here
+                        # TODO: this version info is not used right now. Make use of it later.
+                        jscov_format_version = obj.get("version")
+                        continue
+
+                    try:
+                        for d in process_source_file(
+                            artifact_etl,
+                            counter,
+                            obj,
+                            task_cluster_record
+                        ):
+                            yield value2json(d)
+                    except Exception as e:
+                        Log.warning(
+                            "Error processing test {{test_url}} and source file {{source}} while processing {{key}}",
+                            key=source_key,
+                            test_url=obj.get("testUrl"),
+                            source=obj.get("sourceFile"),
+                            cause=e
+                        )
+
+    def aggregator():
+        aggr_coverage = dict()
+
+        with ZipFile(jsdcov_file) as zipped:
+            for num, zip_name in enumerate(zipped.namelist()):
+                json_stream = ibytes2ilines(zipped.open(zip_name))
+                for source_file_index, obj in enumerate(stream.parse(json_stream, '.', ['.'])):
+                    if please_stop:
+                        Log.error("Shutdown detected. Stopping job ETL.")
+
+                    if source_file_index == 0:
+                        # this is not a jscov object but an object containing the version metadata
+                        # TODO: this metadata should not be here
+                        # TODO: this version info is not used right now. Make use of it later.
+                        jscov_format_version = obj.get("version")
+                        continue
+
+                    obj = wrap(obj)
+
+                    # Collecting coverage information
+                    if obj.sourceFile in aggr_coverage:
+                        covered, uncovered = aggr_coverage[obj.sourceFile]
+                        covered.update(covered.union(set(obj.covered)))
+                        uncovered.update(uncovered.union(set(obj.uncovered)))
```

Comment:
This is also wrong.  It should read  `uncovered = total_covered - covered`

Reply:
I suggest you keep only two statistics: The set of covered lines (`covered`) and the set of all lines (`total_covered`).  These two are well behaved as they are monotonically increasing.  Only when you write a coverage record do you calculate `uncovered` from the other two.

Reply:
Okay, tracking these two statistics does make more sense! I will make this change.